c10/cuda/CUDADeviceAssertion.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/macros/Macros.h>
+
+// Used by preprocessor to convert things to strings
+#define C10_CUDA_DEVICE_ASSERTION_STRINGIFY(x) #x
+#define C10_CUDA_DEVICE_ASSERTION_TOSTRING(x) \
+  C10_CUDA_DEVICE_ASSERTION_STRINGIFY(x)
+#define C10_CUDA_DEVICE_ASSERTION_LINE_STRING \
+  C10_CUDA_DEVICE_ASSERTION_TOSTRING(__LINE__)
+
+namespace c10 {
+namespace cuda {
+
+#ifdef TORCH_USE_CUDA_DSA
+// Copy string from `src` to `dst`
+static __device__ void dstrcpy(char* const dst, const char* const src) {
+  int i = 0;
+  // Copy string from source to destination, ensuring that it
+  // isn't longer than `C10_CUDA_DEVICE_SIDE_MAX_STR_LEN-1`
+  for (; src[i] != '\0' && i < C10_CUDA_DEVICE_SIDE_MAX_STR_LEN - 1; i++) {
+    dst[i] = src[i];
+  }
+  // If the loop terminated before `i` meets the condition below, then
+  // `dst` ends with a null-terminator from `src`. Otherwise, the loop
+  // reached maximum iterations and we add our own null-terminator to
+  // `dst`
+  if (i == C10_CUDA_DEVICE_SIDE_MAX_STR_LEN - 1) {
+    *dst = '\0';
+  }
+}
+
+__device__ __noinline__ void dsa_add_new_assertion_failure(
```

Comment:
since this is noinline, I wonder if we can avoid putting its implementation in a header file